### PR TITLE
Clean up code a bit and add some tests

### PR DESCRIPTION
**Pull request by jslvtr**
Cleaned up the code a bit by extracting functions. I've noticed you've done some cleaning up in the main repo too in the meantime.

I've added a few tests for some of the extracted methods; no integration/system tests for Flask yet.

Would be good to check that the refactoring still has the same functionality. As far as I can tell all should be good to go.

### CODE REVIEW

**Comment by scottrobertson** [comment]
Thanks @jslvtr i will give this a test today! Sorry about the messy code ha...

File: python/main.py [hunk 1]
```diff
@@ -1,139 +1,176 @@
...
 
 from pynYNAB.schema.budget import Transaction
 
-import settings
-import ynab_client
-
 app = Flask('Monzo to YNAB')
 app.config['DEBUG'] = settings.flask_debug
 
-if settings.sentry_dsn:
```
Comment (scottrobertson) [inline python/main.py]: Can we get this added back in somewhere? it's useful for tracking errors in production. Not sure if there is a better way to do it though.

File: python/ynab_client.py [hunk 2]
```diff
@@ -65,6 +67,3 @@ def findPreviousTransaction(payee_name):
         .filter(Transaction.imported_payee==payee_name)\
         .order_by(Transaction.date.desc())\
         .first()
-
-# Sync with YNAB
-sync()
```
Comment (scottrobertson) [inline python/ynab_client.py]: We need the initial sync to happen on process boot as there is a much lower timeout for web requests on heroku.

File: python/main.py [hunk 3]
```diff
@@ -1,139 +1,176 @@
...
 
 @app.route('/')
 def route_index():
     return redirect("https://github.com/scottrobertson/monzo-to-ynab", code=302)
 
+
 @app.route('/webhook', methods=['POST'])
 def route_webhook():
-    expected_delta = 0
+    ynab_client.init()
```
Comment (scottrobertson) [inline python/main.py]: Will this create a new client object every request? It needs to persist across requests to ensure the transactions are cached.

File: python/main.py (hunk 1 above)
Comment (jslvtr) [inline python/main.py]: Ah, totally! Sorry, I removed it because it wasn't in `.env.example` and thought it was not being used.

File: python/main.py (hunk 3 above)
Comment (jslvtr) [inline python/main.py]: Good point, I'll make sure it's a singleton and only inits if it's not already initialized.

File: python/main.py (hunk 1 above)
Comment (scottrobertson) [inline python/main.py]: Oops, sorry. Will get it added there once this PR is merged

File: python/ynab_client.py (hunk 2 above)
Comment (jslvtr) [inline python/ynab_client.py]: Hmmm it's not really a good place for it—`ynab_client` is a library, it shouldn't be doing any processing on import.

I'll think about how to do this.

File: python/main.py (hunk 3 above)
Comment (scottrobertson) [inline python/main.py]: Perfect, thank you. It's fine for small YNAB accounts, but the sync can be quite slow so we need to cache as much as possible :)

File: python/main.py (hunk 1 above)
Comment (jslvtr) [inline python/main.py]: My bad, don't think it's `.env.example`. That may have been something else. I meant `raven` isn't in `requirements.txt` :D

Comment (scottrobertson) [inline python/main.py]: Oh, maybe it was accidently removed when we cleaned up requirements.txt. Feel free to just leave it out for now.

On a side note: i hate pythons dependancy management :D

File: python/ynab_client.py [hunk 4]
```diff
@@ -1,15 +1,20 @@
...
+from sqlalchemy.sql.expression import exists
+
+client = None
+connection = None
+
+def init():
+    global client
+    global connection
+
+    connection = connection or nYnabConnection(settings.ynab_username, settings.ynab_password)
```
Comment (jslvtr) [inline python/ynab_client.py]: Made this a singleton now @scottrobertson

File: python/main.py (hunk 3 above)
Comment (jslvtr) [inline python/main.py]: Done now :+1:

File: python/ynab_client.py (hunk 4 above)
Comment (scottrobertson) [inline python/ynab_client.py]: Perfect. So the next big thing is to allow sync on process boot somehow. We will eventually move this to happen in a background job, but that will always be optional (this needs to be super easy for people to install by default)

File: python/ynab_client.py (hunk 2 above)
Comment (scottrobertson) [inline python/ynab_client.py]: Yeah, it's tricky. Just for some context, Heroku has a 30-second timeout for web requests and a 60-second timeout for process booting. So we moved it there as it works a bit better for bigger accounts.

Comment (jslvtr) [inline python/ynab_client.py]: Do you think something like this could work, then?

```python
if __name__ == '__main__:
    ynab_client.sync()
    app.run(host='0.0.0.0', port=settings.port)
```

Comment (scottrobertson) [inline python/ynab_client.py]: Potentially yeah. I am not sure if the client will be imported at that point though. I would need to experiment with how python loads things.

Comment (jslvtr) [inline python/ynab_client.py]: The client should be imported at that point, all files checked syntactically and compiled, and the app about to run.

Comment (scottrobertson) [inline python/ynab_client.py]: I gave it a try:

```
AttributeError: 'NoneType' object has no attribute 'sync'
```

Comment (scottrobertson) [inline python/ynab_client.py]: Ah, calling `init()` first fixes it

**Comment by scottrobertson** [comment]
```
TypeError: get_payee_details_for_transaction() takes exactly 2 arguments (1 given)
```

File: python/ynab_client.py (hunk 2 above)
Comment (jslvtr) [inline python/ynab_client.py]: Oh, of course—silly me!

Comment (jslvtr) [inline python/ynab_client.py]: I'll move the `init()` there

File: python/main.py [hunk 5]
```diff
@@ -1,139 +1,180 @@
...
+    If a payee with payee_name has been used in the past, we can get their ID and
+    pre-populate category.
+    
+    :param payee_name: The name of the Payee as coming from Monzo.
+    :return: (payee_id, subcategory_id)
+    """
+    previous_transaction = ynab_client.findPreviousTransaction(payee_name)
+    if previous_transaction is not None:
+        settings.log.debug('A previous transaction for the payee %s has been found' % payee_name)
+        return get_payee_details_for_transaction(previous_transaction)
```
Comment (scottrobertson) [inline python/main.py]: We need to pass payee_name here

Comment (jslvtr) [inline python/main.py]: Just fixed

**Comment by jslvtr** [comment]
How do you test this locally by the way?

**Comment by scottrobertson** [comment]
@jslvtr I use `foreman start` which runs the Procfile. But you can just copy the `web` command from the Procfile and run that 👍 

I then have a webhook body from Monzo (you can get this by adding a requestb.in as a webhook URL to Monzo) which I post with Postman.

File: python/main.py [hunk 6]
```diff
@@ -1,139 +1,182 @@
...
+    memo = ''
+    if settings.include_emoji and data['merchant'] and data['merchant'].get('emoji'):
+        memo += ' %s' % data['merchant']['emoji']
+
+    if settings.include_tags and data['merchant'] and data['merchant'].get('metadata', {}).get(
+            'suggested_tags'):
+        memo += ' %s' % data['merchant']['metadata']['suggested_tags']
+
+    # Show the local currency in the notes if this is not in the accounts currency
+    if data['data']['local_currency'] != data['data']['currency']:
```
Comment (scottrobertson) [inline python/main.py]: Can you change this to:

```
    if data['local_currency'] != data['currency']:
        memo += ' (%s %s)' % (data['local_currency'], (abs(data['local_amount']) / 100))
```

We are passing `data['data']` into this method

Comment (jslvtr) [inline python/main.py]: Good catch. Apologies, this was also during merge conflicts—missed it!

File: python/main.py [hunk 7]
```diff
@@ -1,139 +1,182 @@
...
-                        settings.log.debug('We have identified the "%s" category as a good default for this payee' % subcategory.name)
-                        subcategory_id = subcategory.id
-                    else:
-                        settings.log.debug('Split category found, so we will not use that category  for %s' % payee_name)
-                else:
-                    settings.log.debug('A subcategory was not found for the previous transaction for %s' % payee_name)
-            else:
-                settings.log.debug('A previous transaction for the payee %s has not been found' % payee_name)
+    if data.get('type') == 'transaction.created':
+        create_transaction(data['data'], settings, 0)
```
Comment (scottrobertson) [inline python/main.py]: Need a return here

File: python/main.py (hunk 6 above)
Comment (scottrobertson) [inline python/main.py]: Don't worry :) lots of (awesome) changes. Things are bound to get missed!

**Comment by scottrobertson** [comment]
@jslvtr there is a Python version of Foreman that may be more useful for you https://github.com/nickstenning/honcho

So that will run the Procfile. This will be more useful once we add background workers in etc.

File: python/main.py (hunk 7 above)
Comment (jslvtr) [inline python/main.py]: Ah, it'll fail with a no response error I imagine. In the previous code there was a `return ''` at the end.

Something like this:

```python
return jsonify({'message': 'Transaction created in YNAB successfully'})`
```

Seems alright to you?

Comment (scottrobertson) [inline python/main.py]: The `create_transaction` method actually returns `return jsonify(data)` so it should render ok. But having a message like that would be cleaner, agreed.

**Comment by jslvtr** [comment]
That's a good one, thanks for that!

File: python/main.py (hunk 7 above)
Comment (jslvtr) [inline python/main.py]: Good catch, I'll remove `jsonify` from the `create_transaction`. Doesn't make sense there if we're putting a message like this one.

Comment (scottrobertson) [inline python/main.py]: Oh wait, it's because the `create_transaction` may return another message (if it fails). This would render "successful" even if the `create_transaction` returned the following:

```
return jsonify({'error': 'Skipping due to duplicate transaction'})
```

Comment (scottrobertson) [inline python/main.py]: So i would just return 

```
return jsonify({'message': 'Transaction created in YNAB successfully'})`
```

from within the `create_transaction` method, and have the flask method just

```
return create_transaction
```

Comment (jslvtr) [inline python/main.py]: I'll make it so `create_transaction` only returns dicts, that'll be cleaner.

Comment (scottrobertson) [inline python/main.py]: Agreed. Just didn't want to give you too much to do :D

Comment (jslvtr) [inline python/main.py]: Hehe, I'm on board with moving all this logic to a separate `monzo` file or something like that. Plenty more to do!

Comment (scottrobertson) [inline python/main.py]: ha, yes! but we can do that in another PR for sure.

File: python/main.py [hunk 8]
```diff
@@ -1,139 +1,182 @@
...
-            entities_payee_id = ynab_client.getpayee(payee_name).id
-
-        # Create the Transaction
+
+def create_transaction(data, settings, expected_delta):
+    # Sync the account so we get the latest payees
+    ynab_client.sync()
+
+    if data['amount'] == 0:
+        return jsonify({'error': 'Amount is 0'})
```
Comment (scottrobertson) [inline python/main.py]: @jslvtr another one here

File: python/main.py [hunk 9]
```diff
@@ -1,139 +1,182 @@
...
+
+    # Does this account exist?
+    account = ynab_client.getaccount(settings.ynab_account)
+    if not account:
+        return jsonify({'error': 'Account not found'})
```
Comment (scottrobertson) [inline python/main.py]: @jslvtr and here

Comment (jslvtr) [inline python/main.py]: Fixed!

File: python/main.py (hunk 8 above)
Comment (jslvtr) [inline python/main.py]: Fixed!

File: python/main.py [hunk 10]
```diff
@@ -28,23 +28,24 @@ def route_webhook():
...
+        return jsonify({'error': 'Unsupported webhook type'}), 400
 
 
 def create_transaction(data, settings, expected_delta):
     # Sync the account so we get the latest payees
     ynab_client.sync()
 
     if data['amount'] == 0:
-        return jsonify({'error': 'Amount is 0'})
+        return {'error': 'Monzo transaction account is 0.'}, 400
```
Comment (scottrobertson) [inline python/main.py]: I am not sure about this. Returning an error here will cause Monzo to keep retrying. Their webhook is valid, but we just dont want to forward it.

File: python/main.py [hunk 11]
```diff
@@ -100,12 +101,12 @@ def create_transaction(data, settings, expected_delta):
     settings.log.debug('Duplicate detection')
     if ynab_client.containsDuplicate(transaction):
         settings.log.debug('skipping due to duplicate transaction')
-        return {'error': 'Skipping due to duplicate transaction.'}
+        return {'error': 'Tried to add a duplicate transaction.'}, 400
```
Comment (scottrobertson) [inline python/main.py]: Same with this one. We dont want Monzo to retry.

File: python/main.py [hunk 12]
```diff
@@ -28,23 +28,24 @@ def route_webhook():
...
 
     if data['amount'] == 0:
-        return jsonify({'error': 'Amount is 0'})
+        return {'error': 'Monzo transaction account is 0.'}, 400
 
     # Does this account exist?
     account = ynab_client.getaccount(settings.ynab_account)
     if not account:
-        return jsonify({'error': 'Account not found'})
+        return {'error': 'Account {} was not found'.format(settings.ynab_account)}, 400
```
Comment (scottrobertson) [inline python/main.py]: This one is fine being an error

**Comment by scottrobertson** [comment]
@jslvtr other than those error codes, this looks good :)

File: python/main.py (hunk 10 above)
Comment (jslvtr) [inline python/main.py]: Ah, I didn't know of the retry on error. Will make these 200.

File: python/main.py (hunk 11 above)
Comment (jslvtr) [inline python/main.py]: Will make this 200 too.

**Comment by jslvtr** [comment]
@scottrobertson all fixed up! I'll try it out locally when I get the webhook body. Need to go to the shops 😂

**Comment by scottrobertson** [comment]
@jslvtr you can use Stripe :) Charge £1, and the refunds process instantly too :D

**Comment by scottrobertson** [comment]
@jslvtr merged. What is the best way to run these tests?

**Comment by jslvtr** [comment]
Awesome @scottrobertson ! I've been running them with `python -m unittest tests.unit.test_main`.

**Comment by scottrobertson** [comment]
@jslvtr i am getting:

```
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/__main__.py", line 12, in <module>
    main(module=None)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/main.py", line 94, in __init__
    self.parseArgs(argv)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/main.py", line 149, in parseArgs
    self.createTests()
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/main.py", line 158, in createTests
    self.module)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/loader.py", line 130, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/loader.py", line 100, in loadTestsFromName
    parent, obj = obj, getattr(obj, part)
AttributeError: 'module' object has no attribute ''
```

**Comment by jslvtr** [comment]
@scottrobertson that's a bit strange. Seems to work here :/

**Comment by scottrobertson** [comment]
Ah, i copied the trailing `.`. Never mind...
